### PR TITLE
Add flake8 linting, update Makefile

### PR DESCRIPTION
**Pull request by federicotdn**
**Proposed changes**:
This PR adds flake8 to the project. Initially, flake8 will be ran with the `--exit-zero` flag enabled, so that all checks can be fixed incrementally.

Additionally:
- The `.travis.yml` was modified so that it uses `Makefile` recipes. `Makefile` recipes can be now used to lint, format and test the code correctly.
- `pip install black` was removed from `.travis.yml` and added to `requirements-dev.txt`, with a pinned version.

The following `Makefile` recipes were modified:
- `formatter`: Applies black formatting to the code.
- `lint`: Runs flake8 and black checks on the code (does not modify it).
- `types`: Runs pytypes checks on the code.
- `test`: Runs all tests.

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by tabergma** [comment]
@federicotdn When we introduced black we removed the codestyle check. I had a short discussion with @tmbo about that back then (related to https://github.com/RasaHQ/roadmap/issues/436). Why do you think should we reintroduce it again?

File: Makefile [hunk 1]
```diff
@@ -1,21 +1,18 @@
...
+	@echo "    formatter"
+	@echo "        Apply black formatting to code."
 	@echo "    lint"
 	@echo "        Check style with flake8."
+	@echo "    types"
+	@echo "        Check for type errors using pytype."
 	@echo "    test"
 	@echo "        Run py.test"
 	@echo "    check-readme"
 	@echo "        Check if the readme can be converted from md to rst for pypi"
```
Comment (tabergma) [inline Makefile]:
```suggestion
	@echo "        Check if the readme can be converted from md to rst for pypi"
	@echo "    doctest"
	@echo "        Run all doctests embedded in the documentation."
	@echo "    livedocs"
	@echo "        Build the docs locally."
```

**Comment by federicotdn** [comment]
Hey @tabergma,

I think we should use black for code formatting and flake8 for catching common mistakes. They can both be combined [pretty easily](https://black.readthedocs.io/en/stable/the_black_code_style.html?highlight=flake8#line-length).

Adding flake8 would have no impact on end users (as we would only add it as a dev dependency), and it would only add ~3 seconds to the current build times.

By running flake8 once I have already found a couple of moderately serious problems in the code base:

- [Comparing integers by identity](https://github.com/RasaHQ/rasa/blob/master/rasa/core/training/dsl.py#L191)
- [Variable `_train_path` is assigned to but never used](https://github.com/RasaHQ/rasa/blob/master/rasa/cli/train.py#L104), but `train_path` is used later in the function.
- Many unused imports that are potentially making the CLI run a bit slower (I haven't checked this because there are so many of them to remove)

The idea of adding the --exit-zero flag to flake8 was to fix the issues across multiple PRs. There are currently 141 problems found. If we ignore unused imports, variable redefinition and star `*` imports, the number goes down to 7.

**Comment by tabergma** [comment]
Let's add it! Sounds great. Thanks for proposing it!

My concern is that if we use the `--exit-zero` flag, nobody will care about the errors that flake8 detects. I think it is better to enforce those things. But, if you thing it is too much work to fix all the issues, I would recommend that you point this out in Slack, so that everyone knows, that he should look after those issues in their PRs.

**Comment by federicotdn** [comment]
I agree that removing the `--exit-zero` flag is much better than leaving it, to enforce the checks on developers. I think I'll tweak `setup.cfg` to ignore the less serious problems (e.g. unused imports), remove the flag, and fix the remaining problems now. We can later unignore the other problems and continue working on this on other PRs.

File: Makefile [hunk 2]
```diff
@@ -24,22 +25,28 @@ clean:
 	rm -rf build/
 	rm -rf .pytype/
 	rm -rf dist/
-	rm -rf *.egg-info
 	rm -rf docs/_build
 
+formatter:
+	black rasa tests
+
+# TODO: Remove '--exit-zero'
```
Comment (tabergma) [inline Makefile]: Can be removed.

**Comment by erohmensing** [comment]
@federicotdn i came across this when searching for a different travis PR, are there any more flake8 checks you think should be enabled? if so maybe we could create issues + even open that up to the community to help clean up

**Comment by federicotdn** [comment]
@erohmensing I added a couple more back in July: https://github.com/RasaHQ/rasa/pull/4134

Adding new warnings is pretty easy, since we explicitly list the ones that we've disabled (they are all enabled by default). Check the `[flake8]` section in setup.cfg, a good workflow is: 1) delete one of the entries (e.g. E225), 2) run `make lint` and 3) fix warnings. Some of the checks aren't worth removing, in my opinion, because they are not useful for us (for example, imports inside functions).

**Comment by erohmensing** [comment]
Cool, any more in there that you do think would be worth removing?

**Comment by federicotdn** [comment]
I would say `F811` is the most interesting one. Unfortunately it seems to be incorrectly reporting some cases where we import a module twice because it considers `TYPE_CHECKING` to be `True` (because we import some modules for their types and their contents themselves in separate places, but the type checker doesn't care about that).

**Pull request merged**